<a href="https://colab.research.google.com/github/arvindpanda2007/LLM_From_Scratch_solo/blob/tokenizer_v2/tokenization.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!cd

In [ ]:
from google.colab import files

uploaded = files.upload()

Saving the-verdict.txt to the-verdict (2).txt


In [ ]:

with open("the-verdict.txt","r") as f:
    text_R = f.read()

In [ ]:
print("Total number of character:", len(text_R))
print(text_R[:99])

Total number of character: 20479
I HAD always thought Jack Gisburn rather a cheap genius--though a good fellow enough--so it was no 


In [ ]:
import re
text = "Hello, world. This, is a test."
result = re.split(r'(\s)', text)
print(result)
result = [item for item in result if item.strip() != ""]
print(result)

['Hello,', ' ', 'world.', ' ', 'This,', ' ', 'is', ' ', 'a', ' ', 'test.']
['Hello,', 'world.', 'This,', 'is', 'a', 'test.']


In [ ]:
preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)', text_R)
preprocessed = [item.strip() for item in preprocessed if item.strip()]
print(preprocessed)

['I', 'HAD', 'always', 'thought', 'Jack', 'Gisburn', 'rather', 'a', 'cheap', 'genius', '--', 'though', 'a', 'good', 'fellow', 'enough', '--', 'so', 'it', 'was', 'no', 'great', 'surprise', 'to', 'me', 'to', 'hear', 'that', ',', 'in', 'the', 'height', 'of', 'his', 'glory', ',', 'he', 'had', 'dropped', 'his', 'painting', ',', 'married', 'a', 'rich', 'widow', ',', 'and', 'established', 'himself', 'in', 'a', 'villa', 'on', 'the', 'Riviera', '.', '(', 'Though', 'I', 'rather', 'thought', 'it', 'would', 'have', 'been', 'Rome', 'or', 'Florence', '.', ')', '"', 'The', 'height', 'of', 'his', 'glory', '"', '--', 'that', 'was', 'what', 'the', 'women', 'called', 'it', '.', 'I', 'can', 'hear', 'Mrs', '.', 'Gideon', 'Thwing', '--', 'his', 'last', 'Chicago', 'sitter', '--', 'deploring', 'his', 'unaccountable', 'abdication', '.', '"', 'Of', 'course', 'it', "'", 's', 'going', 'to', 'send', 'the', 'value', 'of', 'my', 'picture', "'", 'way', 'up', ';', 'but', 'I', 'don', "'", 't', 'think', 'of', 'that', ',

In [ ]:
all_words = sorted(set(preprocessed))
vocab_size = len(all_words)
print(vocab_size)

1130


In [ ]:
vocab = {token:integer for integer,token in enumerate(all_words)}
for i, item in enumerate(vocab.items()):
    print(item)
    if i >= 50:
        break


('!', 0)
('"', 1)
("'", 2)
('(', 3)
(')', 4)
(',', 5)
('--', 6)
('.', 7)
(':', 8)
(';', 9)
('?', 10)
('A', 11)
('Ah', 12)
('Among', 13)
('And', 14)
('Are', 15)
('Arrt', 16)
('As', 17)
('At', 18)
('Be', 19)
('Begin', 20)
('Burlington', 21)
('But', 22)
('By', 23)
('Carlo', 24)
('Chicago', 25)
('Claude', 26)
('Come', 27)
('Croft', 28)
('Destroyed', 29)
('Devonshire', 30)
('Don', 31)
('Dubarry', 32)
('Emperors', 33)
('Florence', 34)
('For', 35)
('Gallery', 36)
('Gideon', 37)
('Gisburn', 38)
('Gisburns', 39)
('Grafton', 40)
('Greek', 41)
('Grindle', 42)
('Grindles', 43)
('HAD', 44)
('Had', 45)
('Hang', 46)
('Has', 47)
('He', 48)
('Her', 49)
('Hermia', 50)


In [ ]:
from importlib.metadata import version
import tiktoken
print("tiktoken version:", version("tiktoken"))

tiktoken version: 0.12.0


In [ ]:
tokenizer = tiktoken.get_encoding("gpt2")

text = (
    "Hello, do you like tea? <|endoftext|> In the sunlit terraces"
    "of someunknownPlace."
)
integers = tokenizer.encode(text, allowed_special={"<|endoftext|>"})
print(integers)

[15496, 11, 466, 345, 588, 8887, 30, 220, 50256, 554, 262, 4252, 18250, 8812, 2114, 1659, 617, 34680, 27271, 13]


In [ ]:
print(len(text))
print(len(integers))

80
20


In [ ]:
strings = tokenizer.decode(integers)
print(strings)

Hello, do you like tea? <|endoftext|> In the sunlit terracesof someunknownPlace.


BPE tokenizer or byte-pair encoding tokenizer is based upon a sophisticated tokenization scheme which was used to train llms such as ChatGPT

It utilizes the <| endoftext |> token to join two different segments of text

the BPE tokenizer used to train GPT-2,3 and the original model in ChatGPT had a vocabulary size of 50,257 with <| endoftext |> being assigned the largest token id of 50,256

It utilizes an algorithm which breaks down the unknown words not in its vocabulary into a sequence of smaller sub-words or even individual characters to encode it.

Example:


In [ ]:
integers = tokenizer.encode("Kabjngvf :;wfwg")
print(integers)
strings = tokenizer.decode(integers)
print(strings)

[42, 397, 73, 782, 85, 69, 1058, 26, 86, 44482, 70]
Kabjngvf :;wfwg


As we see the "unknown" string was broken down by the tokenizer into a sequence of smaller segments which were effectively decoded too later on

In [ ]:
import tiktoken
tokenizer = tiktoken.get_encoding("gpt2")
with open("the-verdict.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()
enc_text = tokenizer.encode(raw_text)
print(len(enc_text))

5145


In [ ]:
enc_sample = enc_text[50:]

context_size = 5
x = enc_sample[:context_size]
y = enc_sample[1:context_size+1]
print(f"x: {x}")
print(f"y:      {y}")

x: [290, 4920, 2241, 287, 257]
y:      [4920, 2241, 287, 257, 4489]


In [ ]:
for i in range(1, context_size+1):
    context = enc_sample[:i]
    desired = enc_sample[i]
    print(context, "---->", desired)

[290] ----> 4920
[290, 4920] ----> 2241
[290, 4920, 2241] ----> 287
[290, 4920, 2241, 287] ----> 257
[290, 4920, 2241, 287, 257] ----> 4489


In [ ]:
for i in range(1, context_size+1):
    context = enc_sample[:i]
    desired = enc_sample[i]
    print(tokenizer.decode(context), "---->", tokenizer.decode([desired]))

 and ---->  established
 and established ---->  himself
 and established himself ---->  in
 and established himself in ---->  a
 and established himself in a ---->  vill


In [ ]:
from torch.utils.data import Dataset, DataLoader

In [ ]:
import tiktoken
with open("the-verdict.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()
tokenizer = tiktoken.get_encoding("gpt2")

In [ ]:
import torch

In [ ]:
class GPTDataset(Dataset):
    def __init__(self, text, tokenizer, context_size, stride):
        self.input_ids = []
        self.target_ids = []
        tokenized_text = tokenizer.encode(text)
        for i in range(0, len(tokenized_text) - context_size, stride):
            input_id = tokenized_text[i:i+context_size]
            target_id = tokenized_text[i+1:i+context_size+1]
            self.input_ids.append(torch.tensor(input_id))
            self.target_ids.append(torch.tensor(target_id))

    def __len__(self):
        return len(self.input_ids)

    def __getitem__(self, sample_id):
        return self.input_ids[sample_id], self.target_ids[sample_id]

In [ ]:
dataset = GPTDataset(
    raw_text, tokenizer,
    context_size=4,
    stride=1
)

In [ ]:
dataset[0]

([40, 367, 2885, 1464], [367, 2885, 1464, 1807])

In [ ]:
len(dataset)

5141

In [ ]:
raw_text[:100]

'I HAD always thought Jack Gisburn rather a cheap genius--though a good fellow enough--so it was no g'

In [ ]:
tokenizer.decode(dataset[0][0])
tokenizer.decode(dataset[0][1])

' HAD always thought'

In [ ]:
def data_loader(text, batch_size, context_window = 256, stride = 64, shuffle = False,num_workers = 3,drop_last = True):
    tokenizer = tiktoken.get_encoding("gpt2")
    dataset = GPTDataset(
        text, tokenizer,
        context_size=context_window,
        stride=stride
    )
    return DataLoader(dataset, batch_size=batch_size, shuffle=shuffle, num_workers=num_workers, drop_last=drop_last)


In [ ]:
dataloader = data_loader(raw_text, batch_size=4, context_window = 256, stride = 64, shuffle=True, num_workers=3)

/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py:424: UserWarning: This DataLoader will create 3 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  self.check_worker_number_rationality()


In [ ]:
x, y = dataloader.dataset[0]
print(tokenizer.decode(x))
print(tokenizer.decode(y))

I HAD always thought Jack Gisburn rather a cheap genius--though a good fellow enough--so it was no great surprise to me to hear that, in the height of his glory, he had dropped his painting, married a rich widow, and established himself in a villa on the Riviera. (Though I rather thought it would have been Rome or Florence.)

"The height of his glory"--that was what the women called it. I can hear Mrs. Gideon Thwing--his last Chicago sitter--deploring his unaccountable abdication. "Of course it's going to send the value of my picture 'way up; but I don't think of that, Mr. Rickham--the loss to Arrt is all I think of." The word, on Mrs. Thwing's lips, multiplied its _rs_ as though they were reflected in an endless vista of mirrors. And it was not only the Mrs. Thwings who mourned. Had not the exquisite Hermia Croft, at the last Grafton Gallery show, stopped me before Gisburn's "Moon-dancers" to say, with tears in her eyes: "We shall not look upon its like again"?

Well!--
 HAD always th

In [ ]:
for i in range(len(dataloader.dataset)):

    x, y = dataloader.dataset[i]

    print(f"SAMPLE {i}")

    print("INPUT : ", tokenizer.decode(x))
    print("TARGET:", tokenizer.decode(y))

    print("-" * 50)

SAMPLE 0
INPUT :  I HAD always thought Jack Gisburn rather a cheap genius--though a good fellow enough--so it was no great surprise to me to hear that, in the height of his glory, he had dropped his painting, married a rich widow, and established himself in a villa on the Riviera. (Though I rather thought it would have been Rome or Florence.)

"The height of his glory"--that was what the women called it. I can hear Mrs. Gideon Thwing--his last Chicago sitter--deploring his unaccountable abdication. "Of course it's going to send the value of my picture 'way up; but I don't think of that, Mr. Rickham--the loss to Arrt is all I think of." The word, on Mrs. Thwing's lips, multiplied its _rs_ as though they were reflected in an endless vista of mirrors. And it was not only the Mrs. Thwings who mourned. Had not the exquisite Hermia Croft, at the last Grafton Gallery show, stopped me before Gisburn's "Moon-dancers" to say, with tears in her eyes: "We shall not look upon its like again"?

Well

In [ ]:
for batch_x, batch_y in dataloader:

    print(batch_x)

    print(batch_y)

    break

[tensor([ 40, 517, 991, 314]), tensor([ 367,   13, 2045, 2138]), tensor([2885,  383,  546, 1807]), tensor([1464, 3200,  329,  340]), tensor([1807,   30,  257,  561]), tensor([ 3619,  4162, 12854,   423]), tensor([402,  11, 286, 587]), tensor([  271,   314,   884, 10598]), tensor([10899,   550,  3842,   393]), tensor([ 2138,   257,    13, 28537]), tensor([ 257, 3200,  198, 2014]), tensor([7026, 2861,  198,  198]), tensor([15632,  8208,     1,   198]), tensor([  438,   286, 12295,     1]), tensor([2016,  465,  553,  464]), tensor([ 257,    0,  339, 6001]), tensor([922, 314, 531, 286]), tensor([ 5891, 37901, 11589,   465]), tensor([ 1576,   379,    13, 13476]), tensor([438, 262, 198,   1]), tensor([  568, 21978,   198,   438]), tensor([  340, 44896,     1,  5562]), tensor([ 373,   11, 5574,  373]), tensor([ 645,  290, 1660,  644]), tensor([1049, 3088,   12,  262]), tensor([ 5975,   617, 49903,  1466]), tensor([ 284,  286,  438, 1444]), tensor([502, 616, 273, 340]), tensor([  284, 49025,  

The next step after tokenization of raw text sample is to convert them into input-target pairs for training of an llm

for this we will implement two classes
1. Dataset class that inherits __init__ , __len__ , __getitem__ methods from Dataset class in Pytorch library
    a. __init__ method involves taking the text encodings list and initializing input_id list and a target_id list which would later be filled with tensors of inputs and targets respectively
    Note: The tensor arrays input_id and target_id is initially just two lists with tensor elements but due to torch internal mechanism of torch.stack() the entire thing is converted to a tensor

    b. __len__ returns the total number of elements in input_id or target_id or alternatively the total number of input-target pairs

    c. __getitem__ method takes the sample id that is the element number of input_id and returns the corresponding input - target pair

Now after generating the pairs we need to convert this data into batches of input(x) and target(y) stacked together
**Why?**
Because if each pair were dealt with individually then it would really be a lot of waste of computational power since GPUs are built for parallelism
For this we will use the DataLoader class
2. Dataloader class uses the Dataset class previously created and creates batches using Dataloader class in pytorch library


##IMPORTANT PARAMETERS

1. Batch Size: How many samples per batch

2. Context Window: How many tokens can the llm see before predicting next word

3. Shuffle: Shuffling in Dataloader shuffles the order of input and target in their respective batches. It is preferred since it prevents memorization of order

4. Drop_Last: Suppose there is a case where distribution of samples in each input- target pair is same but in the last pair there are very less samples left so distribution becomes lesser then drop last removes it

5. num_workers: num_workers is the number of separate CPU processes PyTorch uses to prepare data in parallel.

6. Stride: stride controls how far the sliding window moves when creating training sequences.

##How these Parameters Affect the Process

*   Larger Batch size improves GPU utilization and faster training but also increases memory usage.
*   Larger Context size improves the models reasoning capability and understanding longer contexts but also increases attention cost.
*   Smaller Stride means more training samples but also means more overlap.
*   Shuffling leads to better convergence and better prevents overfitting of patterns.
*   drop_last = true means there are fixed tensor shapes and stable GPU performance
*   more workers means faster batch preparation and better GPU utilization but increases overhead RAM usage.





##CREATING TOKEN EMBEDDINGS

The last step in preparing input text for LLM training is to convert token IDs to embedding vectors.

The first step is to initialize embedding weights with random values.
Then the model optimizes the vectors based on its discovery of percieved relationships  using backpropagation.

**Now what is backpropagation?**
During the training process the model makes predictions for the next word and then checks the actual result. If they do not match then the model goes back adjusts its internal weights and repeats the procedures. This is called backpropagation mechanism.

The code for the first step is as shown below ☕:


In [ ]:
import torch

#Suppose we have 5 input tokens 2,3,5,1,18

input_token_ids = torch.tensor([2,3,5,1,18])
vocab_size = 19
output_dim = 3

torch.manual_seed(124)

embedding_layer = torch.nn.Embedding(vocab_size, output_dim)

print(embedding_layer)

Embedding(19, 3)


In [ ]:
print(embedding_layer.weight)

Parameter containing:
tensor([[-0.5728,  0.2498,  1.2100],
        [-1.0713,  1.1983,  0.4496],
        [ 0.0454, -1.0395,  0.0552],
        [ 1.3901,  1.4865,  0.3712],
        [-0.0122, -0.2461,  0.8598],
        [-0.4195, -0.0496, -0.9707],
        [-1.2034, -1.5485, -0.2655],
        [ 0.8484,  0.8453, -1.4826],
        [-0.4124,  0.6751, -0.8403],
        [ 1.4388, -0.9206,  0.1130],
        [-0.7559, -0.3033, -1.3482],
        [ 0.7074, -0.8038, -0.2067],
        [ 0.3137,  1.7624, -0.2038],
        [-0.3140,  1.0573,  1.0477],
        [ 0.0862,  0.2267, -0.4979],
        [ 2.3254, -1.2368,  0.0045],
        [-1.7099,  0.5037, -0.1479],
        [ 0.1531, -0.1364, -0.3311],
        [ 1.5150, -0.9470, -1.3890]], requires_grad=True)


In [ ]:
print(embedding_layer.weight.shape)

torch.Size([19, 3])


In [ ]:
print(embedding_layer(input_token_ids))

tensor([[ 0.0454, -1.0395,  0.0552],
        [ 1.3901,  1.4865,  0.3712],
        [-0.4195, -0.0496, -0.9707],
        [-1.0713,  1.1983,  0.4496],
        [ 1.5150, -0.9470, -1.3890]], grad_fn=<EmbeddingBackward0>)


Now embedding_layer is a pytorch object that is not only having a tensor matrix of weights but also has methods and other parameters related to it such as **requires_grad** (which means that backpropagation is allowed on weights).

Now here's a problem

The embedding layer has converted each token ID from the sentence into embeddings but the transformer does not inherently see the sentence based on these embeddings

**Example** : Man bit Dog                                                          
             Dog bit Man

in both cases the same input token ids are put in and the same embeddings come but the position is not known by the transformer

So the solution is also the to additionally include a positional embedding which says its position.

In [ ]:
import torch

vocab_size = 50257
output_dim = 256
embedding_layer = torch.nn.Embedding(vocab_size, output_dim)

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving the-verdict.txt to the-verdict.txt


In [ ]:
with open(r"the-verdict.txt","r") as f:
  raw_text = f.read()

In [ ]:
import tiktoken

tokenizer = tiktoken.get_encoding("gpt2")

In [ ]:
from torch.utils.data import Dataset, DataLoader

In [ ]:
class dataset(Dataset):
  def __init__(self, text, tokenizer, context_size, stride):
    self.input_ids = []
    self.target_ids = []
    tokenized_text = tokenizer.encode(text)
    for i in range(0, len(tokenized_text) - context_size, stride):
      input_id = tokenized_text[i:i+context_size]
      target_id = tokenized_text[i+1:i+context_size+1]
      self.input_ids.append(torch.tensor(input_id))
      self.target_ids.append(torch.tensor(target_id))
    return None

  def __len__(self):
    return len(self.input_ids)

  def __getitem__(self, sample_id):
    return self.input_ids[sample_id], self.target_ids[sample_id]


In [ ]:
dataset_rawtext = dataset(raw_text,tokenizer,4,4)

In [ ]:
dataset_rawtext.__getitem__(3)

(tensor([15632,   438,  2016,   257]), tensor([ 438, 2016,  257,  922]))

In [ ]:
def dataloader(text, batch_size, context_window = 256, stride = 64, shuffle = False,num_workers = 2,drop_last = True):
    tokenizer = tiktoken.get_encoding("gpt2")
    dataset_text = dataset(
        text, tokenizer,
        context_size=context_window,
        stride=stride
    )
    return DataLoader(dataset_text, batch_size=batch_size, shuffle=shuffle, num_workers=num_workers, drop_last=drop_last)

In [ ]:
dataloader_rawtext = dataloader(raw_text, 12, 256, 64, True, 2, True)

In [ ]:
print(dataloader_rawtext.dataset[0])

(tensor([   40,   367,  2885,  1464,  1807,  3619,   402,   271, 10899,  2138,
          257,  7026, 15632,   438,  2016,   257,   922,  5891,  1576,   438,
          568,   340,   373,   645,  1049,  5975,   284,   502,   284,  3285,
          326,    11,   287,   262,  6001,   286,   465, 13476,    11,   339,
          550,  5710,   465, 12036,    11,  6405,   257,  5527, 27075,    11,
          290,  4920,  2241,   287,   257,  4489,    64,   319,   262, 34686,
        41976,    13,   357, 10915,   314,  2138,  1807,   340,   561,   423,
          587, 10598,   393, 28537,  2014,   198,   198,     1,   464,  6001,
          286,   465, 13476,     1,   438,  5562,   373,   644,   262,  1466,
         1444,   340,    13,   314,   460,  3285,  9074,    13, 46606,   536,
         5469,   438, 14363,   938,  4842,  1650,   353,   438,  2934,   489,
         3255,   465, 48422,   540,   450,    67,  3299,    13,   366,  5189,
         1781,   340,   338,  1016,   284,  3758,   262,  1988,

In [ ]:
vocab_size = 50257
embedding_dim = 256

token_embedding = torch.nn.Embedding(
    vocab_size,
    embedding_dim
)

In [ ]:
for inputs, targets in dataloader_rawtext:
  print(inputs.shape)
  break

torch.Size([12, 256])


In [ ]:
embedding_inputs = token_embedding(inputs)

In [ ]:
print(embedding_inputs.shape)

torch.Size([12, 256, 256])


In [ ]:
context_window = 256
pos_embedding_layer = torch.nn.Embedding(
    vocab_size,
    embedding_dim
)

In [ ]:
pos_embeddings = pos_embedding_layer(torch.arange(context_window))

In [ ]:
print(pos_embeddings.shape)

torch.Size([256, 256])


In [ ]:
input_embeddings = embedding_inputs + pos_embeddings

In [ ]:
print(input_embeddings)

tensor([[[ 3.6199e-01, -1.9660e+00, -5.3723e-01,  ...,  1.9754e+00,
           3.6415e-01,  4.0798e-01],
         [ 1.1726e+00,  3.4993e-01,  4.7927e-01,  ...,  2.9513e+00,
          -1.2858e+00, -6.3760e-01],
         [-1.7258e+00, -7.0408e-01, -2.6716e+00,  ...,  1.0782e+00,
          -8.9274e-01, -2.1173e-01],
         ...,
         [-2.7998e+00,  1.3278e+00,  7.6607e-01,  ...,  1.7063e+00,
           2.6774e+00,  5.7839e-02],
         [-1.9717e+00, -6.8792e-02,  8.3240e-01,  ...,  1.2767e-01,
           1.0292e+00, -1.3882e+00],
         [-1.1454e+00, -8.8308e-01, -1.3553e+00,  ...,  1.4350e+00,
          -5.0909e-01, -1.0184e+00]],

        [[ 7.9891e-01, -3.1287e-01, -4.4818e-01,  ...,  2.9505e-01,
          -2.2810e-03,  1.1081e+00],
         [-4.4547e-01,  1.7826e+00, -2.2815e-01,  ...,  2.2633e+00,
          -1.3977e-01,  2.0522e+00],
         [-2.4514e+00, -1.1412e+00, -7.6215e-01,  ..., -1.4816e+00,
           3.0993e-01,  1.2224e+00],
         ...,
         [-9.5344e-01,  1

In [1]:
import torch
inputs = torch.tensor(
  [[0.43, 0.15, 0.89], # Your    (x^1)
   [0.55, 0.87, 0.66], # journey  (x^2)
   [0.57, 0.85, 0.64], # starts  (x^3)
   [0.22, 0.58, 0.33], # with   (x^4)
   [0.77, 0.25, 0.10], # one    (x^5)
   [0.05, 0.80, 0.55]] # step   (x^6)
)

In [2]:
query = inputs[1]
attn_scores_2 = torch.empty(inputs.shape[0])
for i, x_i in enumerate(inputs):
    attn_scores_2[i] = torch.dot(x_i, query)
print(attn_scores_2)

tensor([0.9544, 1.4950, 1.4754, 0.8434, 0.7070, 1.0865])


In [3]:
attn_weights_2_tmp = attn_scores_2 / attn_scores_2.sum()
print("Attention weights:", attn_weights_2_tmp)
print("Sum:", attn_weights_2_tmp.sum())

Attention weights: tensor([0.1455, 0.2278, 0.2249, 0.1285, 0.1077, 0.1656])
Sum: tensor(1.0000)


In [4]:
def softmax_naive(x):
    return torch.exp(x) / torch.exp(x).sum(dim=0)
attn_weights_2_naive = softmax_naive(attn_scores_2)
print("Attention weights:", attn_weights_2_naive)
print("Sum:", attn_weights_2_naive.sum())

Attention weights: tensor([0.1385, 0.2379, 0.2333, 0.1240, 0.1082, 0.1581])
Sum: tensor(1.)


In [6]:
#the softmax function is a better way to normalize attention weights since it gives higher values to important tokens and doesnt just distribute it to irrelevant info.

In [8]:
'''Note that this naive softmax implementation (softmax_naive) may encounter
numerical instability problems, such as overflow and underflow, when dealing with
large or small input values. Therefore, in practice, it’s advisable to use the PyTorch
implementation of softmax, which has been extensively optimized for performance:'''

'Note that this naive softmax implementation (softmax_naive) may encounter\nnumerical instability problems, such as overflow and underflow, when dealing with\nlarge or small input values. Therefore, in practice, it’s advisable to use the PyTorch\nimplementation of softmax, which has been extensively optimized for performance:'

In [9]:
attn_weights_2 = torch.softmax(attn_scores_2, dim=0)
print("Attention weights:", attn_weights_2)
print("Sum:", attn_weights_2.sum())

Attention weights: tensor([0.1385, 0.2379, 0.2333, 0.1240, 0.1082, 0.1581])
Sum: tensor(1.)


In [11]:
#This is the final representation of the context token
query = inputs[1]
context_vec_2 = torch.zeros(query.shape)
for i,x_i in enumerate(inputs):
    context_vec_2 += attn_weights_2[i]*x_i
print(context_vec_2)

tensor([0.4419, 0.6515, 0.5683])
